core/channel-upgrade.go

```diff
@@ -0,0 +1,746 @@
+package core
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"time"
+
+	retry "github.com/avast/retry-go"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	chantypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	"github.com/hyperledger-labs/yui-relayer/log"
+)
+
+type UpgradeState int
+
+const (
+	UPGRADE_STATE_UNINIT UpgradeState = iota
+	UPGRADE_STATE_INIT
+	UPGRADE_STATE_FLUSHING
+	UPGRADE_STATE_FLUSHCOMPLETE
+)
+
+func (state UpgradeState) String() string {
+	switch state {
+	case UPGRADE_STATE_UNINIT:
+		return "UNINIT"
+	case UPGRADE_STATE_INIT:
+		return "INIT"
+	case UPGRADE_STATE_FLUSHING:
+		return "FLUSHING"
+	case UPGRADE_STATE_FLUSHCOMPLETE:
+		return "FLUSHCOMPLETE"
+	default:
+		panic(fmt.Errorf("unexpected UpgradeState: %d", state))
+	}
+}
+
+type UpgradeAction int
+
+const (
+	UPGRADE_ACTION_NONE UpgradeAction = iota
+	UPGRADE_ACTION_TRY
+	UPGRADE_ACTION_ACK
+	UPGRADE_ACTION_CONFIRM
+	UPGRADE_ACTION_OPEN
+	UPGRADE_ACTION_CANCEL
+	UPGRADE_ACTION_CANCEL_FLUSHCOMPLETE
+	UPGRADE_ACTION_TIMEOUT
+)
+
+func (action UpgradeAction) String() string {
+	switch action {
+	case UPGRADE_ACTION_NONE:
+		return "NONE"
+	case UPGRADE_ACTION_TRY:
+		return "TRY"
+	case UPGRADE_ACTION_ACK:
+		return "ACK"
+	case UPGRADE_ACTION_CONFIRM:
+		return "CONFIRM"
+	case UPGRADE_ACTION_OPEN:
+		return "OPEN"
+	case UPGRADE_ACTION_CANCEL:
+		return "CANCEL"
+	case UPGRADE_ACTION_CANCEL_FLUSHCOMPLETE:
+		return "CANCEL_FLUSHCOMPLETE"
+	case UPGRADE_ACTION_TIMEOUT:
+		return "TIMEOUT"
+	default:
+		panic(fmt.Errorf("unexpected UpgradeAction: %d", action))
+	}
+}
+
+// InitChannelUpgrade builds `MsgChannelUpgradeInit` based on the specified UpgradeFields and sends it to the specified chain.
+func InitChannelUpgrade(chain *ProvableChain, upgradeFields chantypes.UpgradeFields) error {
+	logger := GetChannelLogger(chain.Chain)
+	defer logger.TimeTrack(time.Now(), "InitChannelUpgrade")
+
+	addr, err := chain.GetAddress()
+	if err != nil {
+		logger.Error("failed to get address", err)
+		return err
+	}
+
+	msg := chain.Path().ChanUpgradeInit(upgradeFields, addr)
+
+	if _, err := chain.SendMsgs([]sdk.Msg{msg}); err != nil {
+		logger.Error("failed to send MsgChannelUpgradeInit", err)
+		return err
+	} else {
+		logger.Info("successfully initialized channel upgrade")
+	}
+
+	return nil
+}
+
+// ExecuteChannelUpgrade carries out channel upgrade handshake until both chains transition to the OPEN state.
+// This function repeatedly checks the states of both chains and decides the next action.
+func ExecuteChannelUpgrade(pathName string, src, dst *ProvableChain, interval time.Duration, targetSrcState, targetDstState UpgradeState) error {
+	logger := GetChannelPairLogger(src, dst)
+	defer logger.TimeTrack(time.Now(), "ExecuteChannelUpgrade")
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	failures := 0
+	firstCall := true
+	for {
+		<-ticker.C
```

Comment:
Is it the expected behavior for the ticker to block before the first attempt?

Reply:
I've noticed that, but yes, it's certainly not the best code. I will fix it.